### PR TITLE
feat: adjust cart to include line item adjustments

### PR DESCRIPTION
**Pull request by zakariaelas**
### What

- adjust the `CartService` to create/delete/update line item adjustments on the appropriate business events:
   - on `addLineItem`
   - on `updateLineItem`
   - on `applyDiscounts`

### Testing

- check for line item adjustment generation when adding a line item to a cart containing all combination of discount types/allocations
- check for line item adjustment generation when updating a line item to a cart containing all combination of discount types/allocations

### CODE REVIEW

File: packages/medusa/src/api/routes/admin/swaps/index.ts [hunk 1]
```diff
@@ -24,12 +24,14 @@ export default (app) => {
...
   "order",
   "additional_items",
+  "additional_items.adjustments",
   "return_order",
   "fulfillments",
   "payment",
   "shipping_address",
   "shipping_methods",
   "cart",
+  "cart.items.adjustments",
```
Comment (srindom) [inline packages/medusa/src/api/routes/admin/swaps/index.ts:34]: I don't think we can get `cart.items.adjustments` without first getting `cart.items`
```suggestion
  "cart.items",
  "cart.items.adjustments",
```

File: packages/medusa/src/api/routes/store/carts/update-line-item.ts [hunk 2]
```diff
@@ -12,7 +12,6 @@ import { validator } from "../../../../utils/validator"
  * description: "Updates a Line Item if the desired quantity can be fulfilled."
  * parameters:
  *   - (path) id=* {string} The id of the Cart.
- *   - (path) line_id=* {string} The id of the Line Item.
```
Comment (srindom) [inline packages/medusa/src/api/routes/store/carts/update-line-item.ts:15]: Don't think this should be deleted?

File: packages/medusa/src/services/cart.ts [hunk 3]
```diff
@@ -1004,109 +1038,126 @@ class CartService extends BaseService {
...
+      if (discount.usage_limit) {
+        discount.usage_count = discount.usage_count || 0
 
-      if (discount.usage_count >= discount.usage_limit) {
+        if (discount.usage_count >= discount.usage_limit) {
+          throw new MedusaError(
+            MedusaError.Types.NOT_ALLOWED,
+            "Discount has been used maximum allowed times"
+          )
+        }
```
Comment (srindom) [inline packages/medusa/src/services/cart.ts:1072]: nit: we should consider moving this logic to the discount service; maybe we should create a `canApply` method that handles everything from usage conditions to customer permissions to potentially "custom permissions" in the future. 

Don't expect you to include in this PR but noting it here so we can make a ticket for it

Comment (srindom) [reply packages/medusa/src/services/cart.ts:1072]: + date ranges

**Review by srindom — CHANGES_REQUESTED** [review]
Nice work - a couple of small comments

File: packages/medusa/src/api/routes/store/carts/update-line-item.ts (hunk 2 above)
Comment (zakariaelas) [reply packages/medusa/src/api/routes/store/carts/update-line-item.ts]: Whoops! Fixed 😆

File: packages/medusa/src/services/cart.ts (hunk 3 above)
Comment (srindom) [reply packages/medusa/src/services/cart.ts:1072]: By date ranges I mean - the `canApply` method should also take into account if the discount has expired/is scheduled. Sorry for the sub-par explanation 😅

Comment (zakariaelas) [reply packages/medusa/src/services/cart.ts]: yo @srindom! did some changes, lmk what you think!

File: packages/medusa/src/services/cart.ts [hunk 4]
```diff
@@ -570,10 +579,24 @@ class CartService extends BaseService {
...
+
+      // delete all old line item adjustments
+      await this.lineItemAdjustmentService_
+        .withTransaction(manager)
+        .delete({ item_id: result.items.map((item) => item.id) })
+
+      // potentially create/update line item adjustments
+      await this.lineItemAdjustmentService_
+        .withTransaction(manager)
+        .createAdjustments(result)
```
Comment (srindom) [inline packages/medusa/src/services/cart.ts:594]: One area where we might have to double-check if we are good is the case where a swap is created and the swap cart has a return line. The existing adjustments on this line should not be deleted as they will have come with the discounts that were applied when the original order was placed.

We can maybe do `.delete({ item_id:. result.items.filter(i => !i.is_return).map(...) })`?

**Review by srindom — COMMENTED** [review]
Really nice have a small comment that I just want to run by you before we approve

File: packages/medusa/src/services/cart.ts (hunk 4 above)
Comment (zakariaelas) [reply packages/medusa/src/services/cart.ts:594]: well spotted! will fix, thanks!

File: packages/medusa/src/services/cart.ts [hunk 5]
```diff
@@ -457,7 +463,22 @@ class CartService extends BaseService {
...
+        item_id: result.items
+          .filter((i) => !i.is_return)
+          .map((item) => item.id),
+      })
+
+      // adjust existing line item adjustments
+      await this.lineItemAdjustmentService_
+        .withTransaction(manager)
+        .createAdjustments(result)
+
```
Comment (srindom) [inline packages/medusa/src/services/cart.ts:481]: nit: move this into a private CartService method `refreshAdjustments_(cartId)` as this is also used on line 597.

File: integration-tests/api/__tests__/admin/order.js [hunk 6]
```diff
@@ -1555,6 +1555,97 @@ describe("/admin/orders", () => {
...
+        })
+      })
+
+      describe("And given a swap cart", () => {
+        describe("When a line item is added to the swap cart", () => {
+          it("Then should not delete existing return line item adjustments", async () => {
+            const api = useApi()
+
+            const response = await api.post(
+              `/store/carts/disc-swap-cart/line-items`,
```
Comment (srindom) [inline integration-tests/api/__tests__/admin/order.js:1622]: nit (if-minor): use `simpleSwapFactory` or "build" the swap with the API e.g. make the same calls as on lines 1563-1588 above, and then continue with the same cart id. 

I think we would benefit from generally moving away from the "seeder" files and to something more self-contained like the factories/API sequences.

**Review by srindom — APPROVED** [review]
LGTM! Really nice work - have added some small nits if-minor; feel free to merge

**Pull request merged**